### PR TITLE
Use the contract key hasher in kvutils

### PR DESCRIPTION
**Pull request by dajmaki**
Use the Hash.scala contract key hashing to serialize contract keys
in kvutils instead of serializing them as VersionedValues (which isn't guaranteed
to be stable over time).

As discussed with @remyhaemmerle-da, use `Value` instead of `VersionedValue` in `GlobalKey`, and add `hash` to it.

One downside currently is that we need to deserialize a value when computing transaction outputs.
This I expect to be cleaned up when we remove relative contract ids and kvutils will know the transaction outputs up-front when producing the submission message.

TODO:
- [x] Add changelog
- [x] Coordinate the backwards incompatibility of this change with downstream consumers

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 1]
```diff
@@ -228,16 +232,26 @@ object Hash {
...
   def hashPrivateKey(s: String): Hash =
     builder(Purpose.PrivateKey).add(s).build
 
   // This function assumes that key is well typed, i.e. :
-  // 1 - `key.identifier` is the identifier for a template with a key of type τ
-  // 2 - `key.key` is a value of type τ
-  def hashContractKey(key: Node.GlobalKey): Hash =
+  // 1 - `templateId` is the identifier for a template with a key of type τ
+  // 2 - `key` is a value of type τ
+  def hashContractKey(templateId: Identifier, key: Value[Nothing]): Hash =
```
Comment (stefanobaghino-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: `Value[Nothing]`? :thinking:

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 2]
```diff
@@ -343,9 +345,15 @@ object Node {
     }(recorded, isReplayedBy)
 
   /** Useful in various circumstances -- basically this is what a ledger implementation must use as
-    * a key.
+    * a key. The 'hash' should be used when storing the key as value serialization is not guaranteed
+    * to be stable over time (e.g. a new encoding might be introduced).
     */
-  case class GlobalKey(templateId: Identifier, key: VersionedValue[Nothing])
+  case class GlobalKey private (templateId: Identifier, key: Value[Nothing], hash: Hash)
```
Comment (stefanobaghino-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: Not necessary, but a different approach I could suggest is using a `sealed abstract case class`: https://gist.github.com/tpolecat/a5cb0dc9adeacc93f846835ed21c92d2

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala (hunk 1 above)
Comment (dajmaki) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: It must not contain contract ids, hence the `Nothing`. This is following @remyhaemmerle-da 's code.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 2 above)
Comment (dajmaki) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: this language never ceases to amaze me with the amount of footguns it has to offer.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 3]
```diff
@@ -154,7 +157,11 @@ object Hash {
         case Value.ValueText(v) =>
           add(v)
         case Value.ValueContractId(v) =>
-          add(v.coid)
+          purpose match {
+            case Purpose.ContractKey =>
+              sys.error("Hashing of contract id for contract keys is not supported")
+            case _ => add(v.coid)
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: :+1:

**Pull request merged**